src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -381,8 +381,58 @@ private unsafe bool IsValueOfElementType(object value)
         // if this is an array of value classes and that value class has a default constructor
         // then this calls this default constructor on every element in the value class array.
         // otherwise this is a no-op.  Generally this method is called automatically by the compiler
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        public extern void Initialize();
+        public unsafe void Initialize()
+        {
+            MethodTable* pArrayMT = RuntimeHelpers.GetMethodTable(this);
+            TypeHandle thElem = pArrayMT->GetArrayElementTypeHandle();
+            if (thElem.IsTypeDesc)
+            {
+                return;
+            }
+
+            MethodTable* pElemMT = thElem.AsMethodTable();
+            if (!pElemMT->HasDefaultConstructor || !pElemMT->IsValueType)
+            {
+                return;
+            }
+
+            RuntimeType arrayType = (RuntimeType)GetType();
+
+            if (arrayType.GenericCache is not ArrayInitializeCache cache)
+            {
+                cache = new ArrayInitializeCache(arrayType);
+                arrayType.GenericCache = cache;
+            }
+
+            ref byte arrayRef = ref MemoryMarshal.GetArrayDataReference(this);
+            ushort elementSize = pArrayMT->ComponentSize;
+
+            for (int i = 0; i < Length; i++)
+            {
+                cache.InvokeConstructor(ref arrayRef);
+                arrayRef = ref Unsafe.Add(ref arrayRef, elementSize);
+            }
+        }
+
+        private sealed unsafe partial class ArrayInitializeCache
+        {
+            internal delegate*<ref byte, void> ConstructorEntrypoint;
```

Comment:
How about making it `public readonly` and directly invoke it instead of through `InvokeConstructor`?

Reply:
That's what caused the [test failure](https://github.com/dotnet/runtime/pull/77336#issuecomment-1288003955). I can't add a function pointer local to a method on Array or MetadataLoadContext will throw an exception trying to load it.

Reply:
I think it would be better to add a workaround into MetadataLoadContext tests to skip `Initialize` method instead of spreading the workaround around.

Link the workaround to https://github.com/dotnet/runtime/issues/69273.